package/src/activation.ts

```diff
@@ -0,0 +1,470 @@
+import * as sourcegraph from 'sourcegraph'
+import { HandlerArgs, Handler } from './handler'
+import { initLSIF } from './lsif'
+import { impreciseBadge } from './badges'
+import {
+    map,
+    finalize,
+    distinctUntilChanged,
+    shareReplay,
+} from 'rxjs/operators'
+import { Observer, BehaviorSubject, from, Observable, noop } from 'rxjs'
+import { createAbortError } from './abortion'
+
+export type Maybe<T> = { value: T } | undefined
+
+export interface LSIFProviders {
+    hover: (
+        doc: sourcegraph.TextDocument,
+        pos: sourcegraph.Position
+    ) => Promise<Maybe<sourcegraph.Hover | null>>
+
+    definition: (
+        doc: sourcegraph.TextDocument,
+        pos: sourcegraph.Position
+    ) => Promise<Maybe<sourcegraph.Definition>>
+
+    references: (
+        doc: sourcegraph.TextDocument,
+        pos: sourcegraph.Position
+    ) => Promise<Maybe<sourcegraph.Location[] | null>>
+}
+
+export interface SearchProviders {
+    definition: (
+        doc: sourcegraph.TextDocument,
+        pos: sourcegraph.Position
+    ) => Promise<sourcegraph.Definition>
+
+    references: (
+        doc: sourcegraph.TextDocument,
+        pos: sourcegraph.Position
+    ) => Promise<sourcegraph.Location[] | null>
+
+    hover: (
+        doc: sourcegraph.TextDocument,
+        pos: sourcegraph.Position
+    ) => Promise<sourcegraph.Hover | null>
+}
+
+export interface LSPProviders {
+    definition: (
+        doc: sourcegraph.TextDocument,
+        pos: sourcegraph.Position
+    ) => AsyncGenerator<sourcegraph.Definition, void, undefined>
+
+    references: (
+        doc: sourcegraph.TextDocument,
+        pos: sourcegraph.Position,
+        context: sourcegraph.ReferenceContext
+    ) => AsyncGenerator<sourcegraph.Location[] | null, void, undefined>
+
+    hover: (
+        doc: sourcegraph.TextDocument,
+        pos: sourcegraph.Position
+    ) => AsyncGenerator<sourcegraph.Hover | null, void, undefined>
+
+    externalReferences?: ExternalReferenceProvider
+    implementations?: ImplementationsProvider
+}
```

Comment:
These interfaces all look very much the same. Is there a way to avoid the duplication?

Reply:
They're different _for now_ because we'll be adding support for LSIF pagination, etc, and it would be good to have the flexibility in the types.

Reply:
They will be unified if they're obviously similar after doing some cleanup around reference merging (the current behavior of which is very suboptimal: it always uses LSP if a language server is configured, skipping LSIF entirely).